src/robenv/ros/ros.py

```diff
@@ -0,0 +1,109 @@
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+#
+
+from __future__ import annotations
+
+import os
+
+from logging import getLogger
+from pathlib import Path
+from shutil import copy
+
+import requests
+
+from robenv.environment.distro import parse_distro
+from robenv.environment.run_command import run_command
+
+
+_logger = getLogger(__name__)
+
+
+class ROS:
+    def __init__(self, path_or_url: str) -> None:
+        cache_path = Path.home() / ".cache/robenv"
+        xdg_home = os.environ.get("XDG_CACHE_HOME")
+
+        self._archive_path = Path()
+        self._distro_path = Path()
+
+        if xdg_home is not None:
+            cache_path = Path(xdg_home) / "robenv"
+
+        if path_or_url.startswith("http") or path_or_url.endswith("tar.gz"):
+            if "ros2" not in path_or_url:
+                msg = "possible no ros2 archive, get a possible link/file from https://github.com/ros2/ros2/releases"
+                raise ValueError(
+                    msg,
+                )
+            file_name = path_or_url.split("/")[-1]
+            # file_name is somthing like: ros2-humble-20231122-linux-jammy-amd64.tar.bz2
+            file_name_split = file_name.split("-")
+            distro_name = file_name_split[1]
+            _logger.info("%s ", distro_name)
+            self.distro = parse_distro(distro_name)
+            self._archive_path = cache_path / file_name
+            self._distro_path = cache_path / file_name.split(".")[0]
+            cache_path.mkdir(parents=True, exist_ok=True)
+            if "http" in path_or_url:
+                self._download(path_or_url)
+            else:
+                self._copy_to_cache(Path(path_or_url))
+            self._install()
+            self.path = self._find_path()
+        else:
+            _logger.info(" %s ", path_or_url)
+            self.distro = parse_distro(path_or_url.split("/")[3])
+            self.path = Path(path_or_url)
+
+    def _find_path(self) -> Path:
+        search = self._distro_path.glob("**/setup.sh")
+        return next(search).parent
+
+    def _copy_to_cache(self, file_path: Path) -> None:
+        if file_path.exists() and not self._archive_path.exists():
+            copy(file_path, self._archive_path)
+
+    def _download(self, url: str) -> None:
+        if not self._archive_path.exists():
+            _logger.debug("Download %s from %s", self._distro_path.name, url)
+            download = requests.get(url, allow_redirects=True, timeout=60)
+            self._archive_path.write_bytes(download.content)
+            _logger.debug("saved %s at %s", self._distro_path.name, str(self._archive_path))
+
+    def _install(self) -> None:
+        if not self._distro_path.exists():
+            self._distro_path.mkdir(parents=True, exist_ok=True)
+            run_command(command=f"tar xfvj {self._archive_path!s} -C {self._distro_path}", cwd=Path.cwd())
```

Comment:
nitpick

complete outside normal probability but this relies on the program `tar` to be there while we could just use the [`tarfile`](https://docs.python.org/3/library/tarfile.html)-lib and therefore definitely don't need anyone else.